### PR TITLE
Telex

### PR DESCRIPTION
**Pull request by bpcmusic**
Initial pull request for the support of the TELEX modules by the Teletype. Added the 75 new operators supported in the telex.h/c files.

### CODE REVIEW

**Comment by samdoshi** [comment]
Looks fine to me 👍 I haven't check the particulars of `telex.c` though.

If you had a moment you could add a quick note to the `CHANGELOG.md` file, but I think it's missing a bunch of stuff from v1.3 and v1.4 anyway. If you haven't got time, I'll try and get it up to date when I merge my stuff in.

**Comment by bpcmusic** [comment]
Thanks, Sam. Just updated the `changelog.md ` file. Didn't include all 75 new operators, but summarized what was there and linked to the repo where I maintain the list. 

You are right, there is a big gap in the changelog - I didn't go back and fill any of that in.

**Comment by tehn** [comment]
re: changelog, that'd be my fault. i apologize. many i2c repairs were made
in a hurry. i'll go back an update the log.

File: src/ops/telex.c [hunk 1]
```diff
@@ -0,0 +1,654 @@
...
+
+// clang-format on
+
+// telex helpers
+void TXSend(uint8_t model, uint8_t command, uint8_t output, int16_t value,
+            bool set) {
+    // zero-index the output
+    output -= 1;
+    // convert the output to the device and the port
+    uint8_t port = output % 4;
```
Comment (scanner-darkly) [inline src/ops/telex.c]: `output & 3` should be a bit more efficient

File: src/ops/telex.c [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+void TXSet(uint8_t model, uint8_t command, command_state_t *cs) {
+    uint8_t output = cs_pop(cs);
+    int16_t value = cs_pop(cs);
+    TXSend(model, command, output, value, true);
+}
+void TXReceive(uint8_t model, command_state_t *cs, uint8_t mode, bool shift) {
+    // zero-index the output
+    uint8_t input = cs_pop(cs) - 1;
+    // send the port, device and address
+    uint8_t port = input % 4;
```
Comment (scanner-darkly) [inline src/ops/telex.c]: same here

File: src/ops/telex.c [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+    uint8_t device = input >> 2;
+    uint8_t address = model + device;
+    // inputs are numbered 0-7 for each device - shift is for the second half
+    // mode pushes it up so it can read quantized values and note numbers
+    port += (shift ? 4 : 0) + (mode * 8);
```
Comment (scanner-darkly) [inline src/ops/telex.c]: `mode << 3` instead of `mode * 8`?

**Comment by scanner-darkly** [comment]
looks good to me, made a couple of minor suggestions.
will wait for this to be merged before adding `LV.CV` and `CY.CV`, probably makes sense to keep the same version number?

File: src/ops/telex.c (hunk 1 above)
Comment (bpcmusic) [inline src/ops/telex.c]: YES!! Great catch.

File: src/ops/telex.c (hunk 2 above)
Comment (bpcmusic) [inline src/ops/telex.c]: Ditto. ;)

File: src/ops/telex.c (hunk 3 above)
Comment (bpcmusic) [inline src/ops/telex.c]: Another great find! Thank you so much for taking the time to put your eyes to this. Sincerely appreciated.

**Comment by bpcmusic** [comment]
@scanner-darkly loved your suggestions so much I quickly went and made them. Thanks again for taking such a careful look! :)

**Comment by samdoshi** [comment]
@tehn what's the status on this?

I can test on device this weekend 😍

**Comment by tehn** [comment]
looks good to me! i haven't had a moment to test it on hardware, but if you
guys are on it, let's merge

**Pull request merged**